<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>14 Development • AutoSpectral</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="14 Development">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">AutoSpectral</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/01_Full_AutoSpectral_Workflow.html">01 Full Workflow Example</a></li>
    <li><a class="dropdown-item" href="../articles/02_Control_File_example.html">02 Creating the control file</a></li>
    <li><a class="dropdown-item" href="../articles/03_Aurora_example.html">03 Aurora AutoSpectral Example</a></li>
    <li><a class="dropdown-item" href="../articles/04_ID7000_WLS_example.html">04 ID7000 WLS Example</a></li>
    <li><a class="dropdown-item" href="../articles/05_Preparing_Controls.html">05 Preparing Controls</a></li>
    <li><a class="dropdown-item" href="../articles/06_Gating.html">06 Gating</a></li>
    <li><a class="dropdown-item" href="../articles/07_Cleaning.html">07 Control clean-up</a></li>
    <li><a class="dropdown-item" href="../articles/08_Single_Cell_AutoFluorescence.html">08 Single Cell Autofluorescence</a></li>
    <li><a class="dropdown-item" href="../articles/09_Per_Cell_Optimization.html">09 Per Cell Optimization</a></li>
    <li><a class="dropdown-item" href="../articles/10_Reload_AutoSpectral.html">10 Re-running AutoSpectral</a></li>
    <li><a class="dropdown-item" href="../articles/11_Plotting.html">11 Plotting</a></li>
    <li><a class="dropdown-item" href="../articles/12_Speed_It_Up.html">12 Speed It Up (Why So Slow?)</a></li>
    <li><a class="dropdown-item" href="../articles/13_Updates_And_Issues.html">13 Updates and Issues</a></li>
    <li><a class="dropdown-item" href="../articles/14_Development.html">14 Development</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/DrCytometer/AutoSpectral/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>14 Development</h1>
            
            <h4 data-toc-skip class="date">2026-02-11</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/DrCytometer/AutoSpectral/blob/HEAD/vignettes/articles/14_Development.Rmd" class="external-link"><code>vignettes/articles/14_Development.Rmd</code></a></small>
      <div class="d-none name"><code>14_Development.Rmd</code></div>
    </div>

    
    
<p>Here’s some stuff that’s in progress. These changes will show up
first in the development <code>dev</code> branch, which you are welcome
to install and test.</p>
<p>If you would like to contribute to AutoSpectral, these are key areas
I have identified where it has shortcomings or needs improvement.
Suggestions for strategies, or better, code suggestions, would be
appreciated.</p>
<ul>
<li>
<p>Parallelization improvements.</p>
<ul>
<li>The parallelization of functions like
<code>define.flow.control</code> was problematic on Windows due to
memory usage with <code>futures</code> and the tendency of Windows to
usurp threads that R was using.</li>
<li>A new parallel backend using <code>parLapply</code> on Windows or
<code>mclapply</code> on Mac/Linux has been implemented. This should
gracefully fall back to sequential <code>lapply</code> processing if it
runs into any problems when being set-up.</li>
<li>Parallelization of <code><a href="../reference/unmix.autospectral.html">unmix.autospectral()</a></code> in base R can
be ~30% faster for per-cell unmixing if we use <code>foreach</code> with
<code>doParallel</code>, but this introduces at least three new
dependencies and is a minor improvement compared with running via
Rcpp.</li>
<li>I am struggling to produce fast adaptations of AutoSpectral for
MacOS because I do not have access to MacOS. Suggestions would be
appreciated for better implementation of the <code>mclapply</code>
parallelization for Mac. Perhaps more importantly, AutoSpectralRcpp
relies on OpenMP, and OpenMP is difficult to set up and use on Mac, in
my understanding. If it would help, I could produce a version of
AutoSpectralRcpp that just relies on <code>RcppArmadillo</code> without
OpenMP.</li>
<li>Memory management. This is a problem for unmixing large files, I
believe.</li>
</ul>
</li>
<li>
<p>Code clean-up for <code>define.flow.control</code>.</p>
<ul>
<li>This was the first part of the code I wrote, since it is required
for everything downstream. It is, consequently, probably the worst.</li>
<li>I’ve made a little progress on this, but any changes here have
ramifications throughout, so it is slow going.</li>
<li>Currently, we read in all of the FCS data into active memory in R,
gating as we go to reduce data size. It might be better to restructure
this to simply generate gates and retain indices of gated events in each
file. FCS files would need to be read in on the fly for
<code>clean.controls</code> and <code>get.fluorophore.spectra</code>.
This would require a faster FCS reader, such as the
<code>data.table</code>-based approach in Nathan Laniewski’s <a href="https://github.com/nlaniewski/flowstate" class="external-link">flowState</a>.</li>
</ul>
</li>
<li><p>More flexibility in negative/unstained samples.</p></li>
<li><p>Allow multiple controls per fluorophore.</p></li>
<li>
<p>An alternative to the automated gating.</p>
<ul>
<li>This is now in progress.</li>
<li>This will be based on the “Cellular Landmarks” approach from Nathan
Laniewski’s <a href="https://github.com/nlaniewski/flowstate" class="external-link">flowState</a>
</li>
<li>Initial definition of the gate will be based on the location of the
brightest positive events in controls where the marker corresponds to
known lymphocyte or monocyte populations. The code should fall back to
the current gating approach when no known markers are provided.</li>
</ul>
</li>
<li>
<p>Speed up per-cell AF extraction</p>
<ul>
<li>This is now implemented in v1.0.0, figuring out what each cell’s AF
signature is without unmixing each AF possibility on all cells.
Extraction is parallelized via C++ in AutoSpectralRcpp.</li>
</ul>
</li>
<li>
<p>Speed up per-cell fluorophore optimization.</p>
<ul>
<li>This is now implemented in v1.0.0 through strategic screening of
variants for alignment with each cell’s profile/residual.</li>
<li>The pure R version will now benefit from parallelization,
particularly on Mac/Linux.</li>
<li>At this point, I do not expect more major gains in performance. It
has been suggested, however, that GPU acceleration may help. If you have
experience with that and can offer insights, please reach out.</li>
</ul>
</li>
<li>
<p>Fix the issue causing discontinuities.</p>
<ul>
<li>Progress has been made on this in v1.0.0 via the new, sped-up
optimization strategy. The “slow” method should now be fast enough to
fix most of the issue.</li>
<li>Additional improvements include better decontamination of AF from
the spectral variants, which was causing cells to be pushed further away
from the threshold for optimization, and a rduction in low-level noise
introduced into the spectral variants due to fluctuations in electronic
noise or autofluorescence. Any signal in the variant spectrum in a
channel where the optimized single spectrum has less than 5% of the peak
detector signal is now regularized towards the optimized spectrum. This
focuses the variation on the “peaks”, which is what causes the unmixing
errors and spread.</li>
</ul>
</li>
<li>
<p>Better correction of unmixing errors</p>
<ul>
<li>There are several strategies I’m investigating to handling the cases
where the multi-color samples contain obvious errors outside the range
of variation seen in the single-color controls. This is not stuff I’m
going to put online, so get in touch if you’d like to work on this.</li>
</ul>
</li>
<li>
<p>Integration of Poisson IRLS</p>
<ul>
<li>The Poisson IRLS in <code>AutoSpectralRcpp</code> is now working
considerably better and faster. This may allow for integration of the
Poisson optimization after identification of cellular autofluorescence
and fluorophore signatures.</li>
<li>For this to be practical, per-cell fluorophore optimization needs to
be faster (it is now).</li>
<li>This does not appear to provide any additional benefit, so it is no
longer being actively pursued.</li>
</ul>
</li>
</ul>
<p>To install the <code>dev</code> branch:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DrCytometer/AutoSpectral@dev"</span><span class="op">)</span></span></code></pre></div>
<p>To replace this with the (hopefully) stable version, run this:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/remove.packages.html" class="external-link">remove.packages</a></span><span class="op">(</span><span class="st">"AutoSpectral"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"DrCytometer/AutoSpectral"</span><span class="op">)</span></span></code></pre></div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Oliver Burton, Adrian Liston.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
